File: AndroidStudioProjects/HealLotsMobile/app/src/main/java/com/heallots/mobile/ui/activities/RegisterActivity.java
package com.heallots.mobile.ui.activities;

import android.app.DatePickerDialog;
import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.InputType;
import android.text.TextWatcher;
import android.util.Log;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.FrameLayout;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.TextView;
import androidx.appcompat.app.AppCompatActivity;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Locale;
import com.heallots.mobile.R;
import com.heallots.mobile.api.ApiClient;
import com.heallots.mobile.api.ApiService;
import com.heallots.mobile.models.LoginResponse;
import com.heallots.mobile.models.RegisterRequest;
import com.heallots.mobile.storage.TokenManager;
import java.util.HashMap;
import java.util.Map;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class RegisterActivity extends AppCompatActivity {
    private static final String TAG = "RegisterActivity";

    // Form fields
    private EditText nameEditText, emailEditText, phoneEditText, passwordEditText, confirmPasswordEditText;
    private EditText birthdayEditText, addressEditText;
    private Spinner genderSpinner;
    private Button registerButton;
    private TextView loginLink;
    private FrameLayout errorBanner;
    private TextView errorMessage;
    private ImageButton passwordToggleButton, confirmPasswordToggleButton;
    private View passwordStrengthBar;
    private TextView passwordStrengthText;

    // State
    private boolean passwordVisible = false;
    private boolean confirmPasswordVisible = false;
    private boolean isFormatting = false;
    private String selectedBirthday = ""; // Stores birthday as MMDDYYYY for API

    // API & Storage
    private ApiService apiService;
    private TokenManager tokenManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        try {
            setContentView(R.layout.activity_register);

            apiService = ApiClient.getApiService();
            tokenManager = new TokenManager(this);

            // Initialize views
            initializeViews();

            if (!viewsInitialized()) {
                Log.e(TAG, "One or more views are null!");
                finish();
                return;
            }

            setupListeners();
        } catch (Exception e) {
            Log.e(TAG, "Error in onCreate", e);
            finish();
        }
    }

    private void initializeViews() {
        nameEditText = findViewById(R.id.nameEditText);
        emailEditText = findViewById(R.id.emailEditText);
        phoneEditText = findViewById(R.id.phoneEditText);
        birthdayEditText = findViewById(R.id.birthdayEditText);
        genderSpinner = findViewById(R.id.genderSpinner);
        addressEditText = findViewById(R.id.addressEditText);
        passwordEditText = findViewById(R.id.passwordEditText);
        confirmPasswordEditText = findViewById(R.id.confirmPasswordEditText);
        registerButton = findViewById(R.id.registerButton);
        loginLink = findViewById(R.id.loginLink);
        errorBanner = findViewById(R.id.errorBanner);
        errorMessage = findViewById(R.id.errorMessage);
        passwordToggleButton = findViewById(R.id.passwordToggleButton);
        confirmPasswordToggleButton = findViewById(R.id.confirmPasswordToggleButton);
        passwordStrengthBar = findViewById(R.id.passwordStrengthBar);
        passwordStrengthText = findViewById(R.id.passwordStrengthText);
    }

    private boolean viewsInitialized() {
        return nameEditText != null && emailEditText != null && phoneEditText != null &&
                birthdayEditText != null && genderSpinner != null && addressEditText != null &&
                passwordEditText != null && confirmPasswordEditText != null &&
                registerButton != null && loginLink != null && errorBanner != null &&
                errorMessage != null && passwordToggleButton != null && confirmPasswordToggleButton != null &&
                passwordStrengthBar != null && passwordStrengthText != null;
    }

    private void setupListeners() {
        // Register button
        registerButton.setOnClickListener(v -> handleRegister());

        // Login link
        loginLink.setOnClickListener(v -> {
            startActivity(new Intent(RegisterActivity.this, LoginActivity.class));
            finish();
        });

        // Password toggle buttons
        passwordToggleButton.setOnClickListener(v -> togglePasswordVisibility());
        confirmPasswordToggleButton.setOnClickListener(v -> toggleConfirmPasswordVisibility());

        // Phone number auto-formatting
        phoneEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {}

            @Override
            public void afterTextChanged(Editable s) {
                formatPhoneNumber(s);
            }
        });

        // Birthday field: open date picker on click
        birthdayEditText.setInputType(InputType.TYPE_NULL); // Disable keyboard input
        birthdayEditText.setOnClickListener(v -> {
            hideError();
            showDatePicker();
        });
        birthdayEditText.setOnFocusChangeListener((v, hasFocus) -> {
            if (hasFocus) {
                hideError();
                showDatePicker();
            }
        });

        // Password strength indicator
        passwordEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {}

            @Override
            public void afterTextChanged(Editable s) {
                updatePasswordStrength();
            }
        });

        // Clear error on field focus
        nameEditText.setOnFocusChangeListener((v, hasFocus) -> { if (hasFocus) hideError(); });
        emailEditText.setOnFocusChangeListener((v, hasFocus) -> { if (hasFocus) hideError(); });
        phoneEditText.setOnFocusChangeListener((v, hasFocus) -> { if (hasFocus) hideError(); });
        passwordEditText.setOnFocusChangeListener((v, hasFocus) -> { if (hasFocus) hideError(); });
        confirmPasswordEditText.setOnFocusChangeListener((v, hasFocus) -> { if (hasFocus) hideError(); });

        // Setup Gender Spinner
        String[] genderOptions = {"Select Gender", "Male", "Female", "Other"};
        ArrayAdapter<String> genderAdapter = new ArrayAdapter<>(this,
                android.R.layout.simple_spinner_item, genderOptions);
        genderAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        genderSpinner.setAdapter(genderAdapter);
    }

    private void formatPhoneNumber(Editable editable) {
        if (isFormatting) return;
        isFormatting = true;

        String text = editable.toString().replaceAll("[^0-9]", "");

        if (text.length() > 11) {
            text = text.substring(0, 11);
        }

        String formatted = "";
        if (text.length() > 0) {
            if (text.length() <= 4) {
                formatted = text;
            } else if (text.length() <= 7) {
                formatted = text.substring(0, 4) + " " + text.substring(4);
            } else {
                formatted = text.substring(0, 4) + " " + text.substring(4, 7) + " " + text.substring(7);
            }
        }

        if (!editable.toString().equals(formatted)) {
            editable.clear();
            editable.append(formatted);
        }

        isFormatting = false;
    }

    /**
     * Show DatePickerDialog for birthday selection
     */
    private void showDatePicker() {
        Calendar calendar = Calendar.getInstance();
        
        DatePickerDialog datePickerDialog = new DatePickerDialog(
            this,
            (view, year, month, dayOfMonth) -> {
                // Store in MMDDYYYY format for API
                String monthStr = String.format(Locale.US, "%02d", month + 1); // month is 0-indexed
                String dayStr = String.format(Locale.US, "%02d", dayOfMonth);
                String yearStr = String.format(Locale.US, "%04d", year);
                selectedBirthday = monthStr + dayStr + yearStr;
                
                // Display in "Month DD, YYYY" format
                Calendar selectedDate = Calendar.getInstance();
                selectedDate.set(year, month, dayOfMonth);
                SimpleDateFormat displayFormat = new SimpleDateFormat("MMMM dd, yyyy", Locale.US);
                birthdayEditText.setText(displayFormat.format(selectedDate.getTime()));
            },
            calendar.get(Calendar.YEAR),
            calendar.get(Calendar.MONTH),
            calendar.get(Calendar.DAY_OF_MONTH)
        );
        
        datePickerDialog.show();
    }

    private void updatePasswordStrength() {
        String password = passwordEditText.getText().toString().trim();
        int length = password.length();
        boolean hasUppercase = password.matches(".*[A-Z].*");
        boolean hasNumber = password.matches(".*\\d.*");

        int strengthColor;
        String strengthText;
        float fillPercent;

        if (length < 6) {
            strengthColor = getResources().getColor(R.color.strength_too_short);
            strengthText = "Too short";
            fillPercent = 0.25f;
        } else if (length >= 6 && length <= 8 && (!hasUppercase || !hasNumber)) {
            strengthColor = getResources().getColor(R.color.strength_weak);
            strengthText = "Weak";
            fillPercent = 0.50f;
        } else if (length > 8 && (!hasUppercase || !hasNumber)) {
            strengthColor = getResources().getColor(R.color.strength_fair);
            strengthText = "Fair";
            fillPercent = 0.70f;
        } else {
            strengthColor = getResources().getColor(R.color.strength_strong);
            strengthText = "Strong";
            fillPercent = 1.0f;
        }

        // Update strength bar width using LinearLayout.LayoutParams
        try {
            LinearLayout.LayoutParams params = (LinearLayout.LayoutParams) passwordStrengthBar.getLayoutParams();
            if (params != null) {
                android.view.ViewGroup parent = (android.view.ViewGroup) passwordStrengthBar.getParent();
                int parentWidth = parent != null ? parent.getWidth() : 0;

                if (parentWidth <= 0) {
                    int screenWidth = getResources().getDisplayMetrics().widthPixels;
                    parentWidth = (int) (screenWidth * 0.8);
                }

                int barWidth = (int) (parentWidth * fillPercent);
                params.width = barWidth;
                passwordStrengthBar.setLayoutParams(params);
            }
        } catch (Exception e) {
            Log.e(TAG, "Error updating password strength bar: " + e.getMessage());
        }

        passwordStrengthBar.setBackgroundColor(strengthColor);
        passwordStrengthText.setText(strengthText);
        passwordStrengthText.setTextColor(strengthColor);
    }

    private void togglePasswordVisibility() {
        passwordVisible = !passwordVisible;
        if (passwordVisible) {
            passwordEditText.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD);
        } else {
            passwordEditText.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_PASSWORD);
        }
        passwordEditText.setSelection(passwordEditText.getText().length());
    }

    private void toggleConfirmPasswordVisibility() {
        confirmPasswordVisible = !confirmPasswordVisible;
        if (confirmPasswordVisible) {
            confirmPasswordEditText.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD);
        } else {
            confirmPasswordEditText.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_PASSWORD);
        }
        confirmPasswordEditText.setSelection(confirmPasswordEditText.getText().length());
    }

    private void showError(String message) {
        errorMessage.setText(message);
        errorBanner.setVisibility(View.VISIBLE);
    }

    private void hideError() {
        errorBanner.setVisibility(View.GONE);
    }

    private void hideKeyboard() {
        InputMethodManager imm = (InputMethodManager) getSystemService(INPUT_METHOD_SERVICE);
        View view = getCurrentFocus();
        if (view != null && imm != null) {
            imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
        }
    }

    private void handleRegister() {
        String name = nameEditText.getText().toString().trim();
        String email = emailEditText.getText().toString().trim();
        String phone = phoneEditText.getText().toString().trim().replaceAll(" ", "");
        String birthday = selectedBirthday; // Use the stored MMDDYYYY format
        String gender = genderSpinner.getSelectedItem().toString();
        String address = addressEditText.getText().toString().trim();
        String password = passwordEditText.getText().toString().trim();
        String confirmPassword = confirmPasswordEditText.getText().toString().trim();

        if (!validateInputs(name, email, phone, birthday, gender, address, password, confirmPassword)) {
            return;
        }

        hideKeyboard();
        registerButton.setEnabled(false);
        registerButton.setText("Creating Account...");

        RegisterRequest request = new RegisterRequest(name, email, phone, birthday, gender, address, password);

        apiService.register(request).enqueue(new Callback<LoginResponse>() {
            @Override
            public void onResponse(Call<LoginResponse> call, Response<LoginResponse> response) {
                if (response.isSuccessful() && response.body() != null) {
                    LoginResponse registerResponse = response.body();
                    tokenManager.saveToken(registerResponse.getToken());
                    tokenManager.saveUser(registerResponse.getUser());

                    Log.d(TAG, "Registration successful");
                    startActivity(new Intent(RegisterActivity.this, DashboardActivity.class));
                    finish();
                } else {
                    String errorMsg = "Registration failed";
                    if (response.code() == 409) {
                        errorMsg = "Email already registered";
                    }
                    showError(errorMsg);
                    Log.e(TAG, "Registration error: " + response.code());
                    registerButton.setEnabled(true);
                    registerButton.setText("Create Account");
                }
            }

            @Override
            public void onFailure(Call<LoginResponse> call, Throwable t) {
                showError("Error: " + (t.getMessage() != null ? t.getMessage() : "Unknown error"));
                Log.e(TAG, "Registration failed", t);
                registerButton.setEnabled(true);
                registerButton.setText("Create Account");
            }
        });
    }

    private boolean validateInputs(String name, String email, String phone, String birthday,
                                   String gender, String address, String password, String confirmPassword) {
        if (name.isEmpty()) {
            showError("Full name is required");
            return false;
        }
        if (name.length() < 2) {
            showError("Name must be at least 2 characters");
            return false;
        }
        if (email.isEmpty()) {
            showError("Email address is required");
            return false;
        }
        if (!android.util.Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
            showError("Invalid email format (e.g., maria@example.com)");
            return false;
        }
        if (phone.isEmpty() || phone.length() < 10) {
            showError("Phone number must be at least 10 digits");
            return false;
        }
        if (birthday.isEmpty() || birthday.length() != 8) {
            showError("Birthday must be complete (MM/DD/YYYY)");
            return false;
        }
        if (gender.equals("Select Gender")) {
            showError("Please select a gender");
            return false;
        }
        if (address.isEmpty()) {
            showError("Address is required");
            return false;
        }
        if (address.length() < 5) {
            showError("Address must be at least 5 characters");
            return false;
        }
        if (password.isEmpty()) {
            showError("Password is required");
            return false;
        }
        if (password.length() < 6) {
            showError("Password must be at least 6 characters");
            return false;
        }
        if (confirmPassword.isEmpty()) {
            showError("Please confirm your password");
            return false;
        }
        if (!password.equals(confirmPassword)) {
            showError("Passwords do not match");
            return false;
        }
        return true;
    }
}

File: AndroidStudioProjects/HealLotsMobile/app/src/main/java/com/heallots/mobile/ui/activities/RegisterActivity.kt
package com.heallots.mobile.features.auth.register

import android.app.DatePickerDialog
import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.InputType
import android.text.TextWatcher
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import android.widget.ArrayAdapter
import android.widget.Button
import android.widget.EditText
import android.widget.FrameLayout
import android.widget.ImageButton
import android.widget.LinearLayout
import android.widget.Spinner
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.heallots.mobile.R
import com.heallots.mobile.api.ApiClient
import com.heallots.mobile.features.appointments.dashboard.DashboardActivity
import com.heallots.mobile.features.auth.login.LoginActivity
import com.heallots.mobile.storage.TokenManager
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Locale

class RegisterActivity : AppCompatActivity(), RegisterContract.View {
    private lateinit var nameEditText: EditText
    private lateinit var emailEditText: EditText
    private lateinit var phoneEditText: EditText
    private lateinit var passwordEditText: EditText
    private lateinit var confirmPasswordEditText: EditText
    private lateinit var birthdayEditText: EditText
    private lateinit var addressEditText: EditText
    private lateinit var genderSpinner: Spinner
    private lateinit var registerButton: Button
    private lateinit var loginLink: TextView
    private lateinit var errorBanner: FrameLayout
    private lateinit var errorMessage: TextView
    private lateinit var passwordToggleButton: ImageButton
    private lateinit var confirmPasswordToggleButton: ImageButton
    private lateinit var passwordStrengthBar: View
    private lateinit var passwordStrengthText: TextView

    private var passwordVisible = false
    private var confirmPasswordVisible = false
    private var isFormatting = false
    private var selectedBirthday = ""

    private lateinit var presenter: RegisterContract.Presenter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        try {
            setContentView(R.layout.activity_register)

            presenter = RegisterPresenter(
                view = this,
                repository = RegisterRepository(
                    apiService = ApiClient.getApiService(),
                    tokenManager = TokenManager(this)
                )
            )

            initializeViews()
            setupListeners()
        } catch (e: Exception) {
            Log.e(TAG, "Error in onCreate", e)
            finish()
        }
    }

    private fun initializeViews() {
        nameEditText = findViewById(R.id.nameEditText)
        emailEditText = findViewById(R.id.emailEditText)
        phoneEditText = findViewById(R.id.phoneEditText)
        birthdayEditText = findViewById(R.id.birthdayEditText)
        genderSpinner = findViewById(R.id.genderSpinner)
        addressEditText = findViewById(R.id.addressEditText)
        passwordEditText = findViewById(R.id.passwordEditText)
        confirmPasswordEditText = findViewById(R.id.confirmPasswordEditText)
        registerButton = findViewById(R.id.registerButton)
        loginLink = findViewById(R.id.loginLink)
        errorBanner = findViewById(R.id.errorBanner)
        errorMessage = findViewById(R.id.errorMessage)
        passwordToggleButton = findViewById(R.id.passwordToggleButton)
        confirmPasswordToggleButton = findViewById(R.id.confirmPasswordToggleButton)
        passwordStrengthBar = findViewById(R.id.passwordStrengthBar)
        passwordStrengthText = findViewById(R.id.passwordStrengthText)
    }

    override fun onDestroy() {
        presenter.onDestroy()
        super.onDestroy()
    }

    private fun setupListeners() {
        registerButton.setOnClickListener { handleRegister() }

        loginLink.setOnClickListener { presenter.onLoginClicked() }

        passwordToggleButton.setOnClickListener { togglePasswordVisibility() }
        confirmPasswordToggleButton.setOnClickListener { toggleConfirmPasswordVisibility() }

        phoneEditText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) = Unit
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) = Unit
            override fun afterTextChanged(s: Editable) {
                formatPhoneNumber(s)
            }
        })

        birthdayEditText.inputType = InputType.TYPE_NULL
        birthdayEditText.setOnClickListener {
            hideError()
            showDatePicker()
        }
        birthdayEditText.setOnFocusChangeListener { _, hasFocus ->
            if (hasFocus) {
                hideError()
                showDatePicker()
            }
        }

        passwordEditText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) = Unit
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) = Unit
            override fun afterTextChanged(s: Editable) {
                updatePasswordStrength()
            }
        })

        nameEditText.setOnFocusChangeListener { _, hasFocus -> if (hasFocus) hideError() }
        emailEditText.setOnFocusChangeListener { _, hasFocus -> if (hasFocus) hideError() }
        phoneEditText.setOnFocusChangeListener { _, hasFocus -> if (hasFocus) hideError() }
        passwordEditText.setOnFocusChangeListener { _, hasFocus -> if (hasFocus) hideError() }
        confirmPasswordEditText.setOnFocusChangeListener { _, hasFocus -> if (hasFocus) hideError() }

        val genderOptions = arrayOf("Select Gender", "Male", "Female", "Other")
        val genderAdapter = ArrayAdapter(this, android.R.layout.simple_spinner_item, genderOptions)
        genderAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        genderSpinner.adapter = genderAdapter
    }

    private fun formatPhoneNumber(editable: Editable) {
        if (isFormatting) return
        isFormatting = true

        var text = editable.toString().replace(Regex("[^0-9]"), "")
        if (text.length > 11) {
            text = text.substring(0, 11)
        }

        val formatted = when {
            text.isEmpty() -> ""
            text.length <= 4 -> text
            text.length <= 7 -> "${text.substring(0, 4)} ${text.substring(4)}"
            else -> "${text.substring(0, 4)} ${text.substring(4, 7)} ${text.substring(7)}"
        }

        if (editable.toString() != formatted) {
            editable.clear()
            editable.append(formatted)
        }

        isFormatting = false
    }

    private fun showDatePicker() {
        val calendar = Calendar.getInstance()

        val datePickerDialog = DatePickerDialog(
            this,
            { _, year, month, dayOfMonth ->
                val monthStr = String.format(Locale.US, "%02d", month + 1)
                val dayStr = String.format(Locale.US, "%02d", dayOfMonth)
                val yearStr = String.format(Locale.US, "%04d", year)
                selectedBirthday = monthStr + dayStr + yearStr

                val selectedDate = Calendar.getInstance()
                selectedDate.set(year, month, dayOfMonth)
                val displayFormat = SimpleDateFormat("MMMM dd, yyyy", Locale.US)
                birthdayEditText.setText(displayFormat.format(selectedDate.time))
            },
            calendar.get(Calendar.YEAR),
            calendar.get(Calendar.MONTH),
            calendar.get(Calendar.DAY_OF_MONTH)
        )

        datePickerDialog.show()
    }

    private fun updatePasswordStrength() {
        val password = passwordEditText.text.toString().trim()
        val length = password.length
        val hasUppercase = password.matches(Regex(".*[A-Z].*"))
        val hasNumber = password.matches(Regex(".*\\d.*"))

        val strengthColor: Int
        val strengthText: String
        val fillPercent: Float

        when {
            length < 6 -> {
                strengthColor = resources.getColor(R.color.strength_too_short)
                strengthText = "Too short"
                fillPercent = 0.25f
            }
            length in 6..8 && (!hasUppercase || !hasNumber) -> {
                strengthColor = resources.getColor(R.color.strength_weak)
                strengthText = "Weak"
                fillPercent = 0.50f
            }
            length > 8 && (!hasUppercase || !hasNumber) -> {
                strengthColor = resources.getColor(R.color.strength_fair)
                strengthText = "Fair"
                fillPercent = 0.70f
            }
            else -> {
                strengthColor = resources.getColor(R.color.strength_strong)
                strengthText = "Strong"
                fillPercent = 1.0f
            }
        }

        try {
            val params = passwordStrengthBar.layoutParams as? LinearLayout.LayoutParams
            if (params != null) {
                val parent = passwordStrengthBar.parent as? ViewGroup
                var parentWidth = parent?.width ?: 0

                if (parentWidth <= 0) {
                    val screenWidth = resources.displayMetrics.widthPixels
                    parentWidth = (screenWidth * 0.8f).toInt()
                }

                params.width = (parentWidth * fillPercent).toInt()
                passwordStrengthBar.layoutParams = params
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error updating password strength bar: ${e.message}")
        }

        passwordStrengthBar.setBackgroundColor(strengthColor)
        passwordStrengthText.text = strengthText
        passwordStrengthText.setTextColor(strengthColor)
    }

    private fun togglePasswordVisibility() {
        passwordVisible = !passwordVisible
        passwordEditText.inputType = if (passwordVisible) {
            InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD
        } else {
            InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_PASSWORD
        }
        passwordEditText.setSelection(passwordEditText.text.length)
    }

    private fun toggleConfirmPasswordVisibility() {
        confirmPasswordVisible = !confirmPasswordVisible
        confirmPasswordEditText.inputType = if (confirmPasswordVisible) {
            InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD
        } else {
            InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_PASSWORD
        }
        confirmPasswordEditText.setSelection(confirmPasswordEditText.text.length)
    }

    override fun showError(message: String) {
        errorMessage.text = message
        errorBanner.visibility = View.VISIBLE
    }

    override fun hideError() {
        errorBanner.visibility = View.GONE
    }

    private fun hideKeyboard() {
        val imm = getSystemService(INPUT_METHOD_SERVICE) as? InputMethodManager
        val view = currentFocus
        if (view != null && imm != null) {
            imm.hideSoftInputFromWindow(view.windowToken, 0)
        }
    }

    private fun handleRegister() {
        hideKeyboard()
        presenter.onRegisterClicked(
            RegisterForm(
                name = nameEditText.text.toString(),
                email = emailEditText.text.toString(),
                phone = phoneEditText.text.toString(),
                birthday = selectedBirthday,
                gender = genderSpinner.selectedItem.toString(),
                address = addressEditText.text.toString(),
                password = passwordEditText.text.toString(),
                confirmPassword = confirmPasswordEditText.text.toString()
            )
        )
    }

    override fun showLoading() {
        registerButton.isEnabled = false
        registerButton.text = "Creating Account..."
    }

    override fun hideLoading() {
        registerButton.isEnabled = true
        registerButton.text = "Create Account"
    }

    override fun navigateToLogin() {
        startActivity(Intent(this, LoginActivity::class.java))
        finish()
    }

    override fun navigateToDashboard() {
        Log.d(TAG, "Registration successful")
        startActivity(Intent(this, DashboardActivity::class.java))
        finish()
    }

    companion object {
        private const val TAG = "RegisterActivity"
    }
}
